lib/ansible/module_utils/dimensiondata.py

```diff
@@ -0,0 +1,393 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2016 Dimension Data
+#
+# This module is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This software is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this software.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Authors:
+#   - Aimon Bustardo <aimon.bustardo@dimensiondata.com>
+#   - Mark Maglana   <mmaglana@gmail.com>
+#
+# Common methods to be used by versious module components
+import os
+import ConfigParser
+from os.path import expanduser
+from uuid import UUID
+
+try:
+    from libcloud.common.dimensiondata import \
+        API_ENDPOINTS, DimensionDataAPIException
+    HAS_LIBCLOUD = True
+except ImportError:
+    HAS_LIBCLOUD = False
+
+
+# Custom Exceptions
+
+class LibcloudNotFound(Exception):
+    pass
+
+
+class MissingCredentialsError(Exception):
+    pass
+
+
+class UnknownNetworkError(Exception):
+    pass
+
+
+class UnknownVLANError(Exception):
+    pass
+
+
+def check_libcloud_or_fail():
+    """
+    Checks if libcloud is installed and fails if not
+    """
+    if not HAS_LIBCLOUD:
+        raise LibcloudNotFound("apache-libcloud is required.")
+
+
+def get_credentials():
+    """
+    This method will get user_id and key from environemnt or dot file.
+    Environment takes priority.
+
+    To set in environment:
+
+        export DIDATA_USER='myusername'
+        export DIDATA_PASSWORD='mypassword'
+
+    To set in dot file place a file at ~/.dimensiondata with
+    the following contents:
+
+        [dimensiondatacloud]
+        DIDATA_USER: myusername
+        DIDATA_PASSWORD: mypassword
+    """
+    check_libcloud_or_fail()
+
+    # Attempt to grab from environment
+    user_id = os.environ.get('DIDATA_USER', None)
+    key = os.environ.get('DIDATA_PASSWORD', None)
+
+    # Environment failed try dot file
+    if not user_id or not key:
+        home = expanduser('~')
+        config = ConfigParser.RawConfigParser()
+        config.read("%s/.dimensiondata" % home)
+        try:
+            user_id = config.get("dimensiondatacloud", "DIDATA_USER")
+            key = config.get("dimensiondatacloud", "DIDATA_PASSWORD")
+        except ConfigParser.NoSectionError, ConfigParser.NoOptionError:
+            pass
+
+    # One or more credentials not found. Function can't recover from this
+    # so it has to raise an error instead of fail silently.
+    if not user_id:
+        raise MissingCredentialsError("Dimension Data user id not found")
+    elif not key:
+        raise MissingCredentialsError("Dimension Data key not found")
```

Comment:
@tintoy, yes that's about what I'm talking about.  I'd probably put all that logic into a single function, though.  So you'd call get_credentials(module) and it would first look in the module parameters, then the environment, and finally in the dot files (returning the first set that it found).